message_ix/macro.py

```diff
@@ -1,61 +1,494 @@
+import collections
+import os
+
+import numpy as np
+import pandas as pd
+
+from functools import lru_cache
+
+from ixmp.utils import logger
+
+#
+# TODOS:
+#
+# 1) all demands/prices assumed to be on USEFUL level, need to extend this
+#    to support others
+#
 
 MACRO_INIT = {
-    'sets': [
-        ['sector', ],
-        ['mapping_macro_sector', ['sector', 'commodity', 'level', ], ],
-    ],
-    'pars': [
-        ['demand_MESSAGE', ['node', 'sector', 'year', ], ],
-        ['price_MESSAGE', ['node', 'sector', 'year', ], ],
-        ['cost_MESSAGE', ['node', 'year', ], ],
-        ['gdp_calibrate', ['node', 'year', ], ],
-        ['MERtoPPP', ['node', 'year', ], ],
-        ['kgdp', ['node', ], ],
-        ['kpvs', ['node', ], ],
-        ['depr', ['node', ], ],
-        ['drate', ['node', ], ],
-        ['esub', ['node', ], ],
-        ['lotol', ['node', ], ],
-        ['p_ref', ['node', 'sector', ], ],
-        ['lakl', ['node', ], ],
-        ['prfconst', ['node', 'sector', ], ],
-        ['grow', ['node', 'year', ], ],
-        ['aeei', ['node', 'sector', 'year', ], ],
-    ],
-    'vars': [
-        ['DEMAND', ['node', 'commodity', 'level', 'year', 'time', ], ],
-        ['PRICE', ['node', 'commodity', 'level', 'year', 'time', ], ],
-        ['COST_NODAL', ['node', 'year', ], ],
-        ['COST_NODAL_NET', ['node', 'year', ], ],
-        ['GDP', ['node', 'year', ], ],
-        ['I', ['node', 'year', ], ],
-        ['C', ['node', 'year', ], ],
-        ['K', ['node', 'year', ], ],
-        ['KN', ['node', 'year', ], ],
-        ['Y', ['node', 'year', ], ],
-        ['YN', ['node', 'year', ], ],
-        ['EC', ['node', 'year', ], ],
-        ['UTILITY', ],
-        ['PHYSENE', ['node', 'sector', 'year', ], ],
-        ['PRODENE', ['node', 'sector', 'year', ], ],
-        ['NEWENE', ['node', 'sector', 'year', ], ],
-        ['EC', ['node', 'year', ], ],
-        ['grow_calibrate', ['node', 'year', ], ],
-        ['aeei_calibrate', ['node', 'sector', 'year', ], ],
-    ],
-    'equs': [
-        ['COST_ACCOUNTING_NODAL', ['node', 'year', ], ]
-    ],
+    'sets': {
+        'sector': None,
+        'mapping_macro_sector': ['sector', 'commodity', 'level', ],
+    },
+    'pars': {
+        'demand_MESSAGE': {
+            'idx': ['node', 'sector', 'year', ],
+            'unit': 'GWa',
```

Comment:
As for any MESSAGEix model, units in any MACRO model are not defined ex-ante for each model instance, but can in principle be freely chosen. While this set of units might be the right ones in the most frequently used models, I don't think they should be hard coded into this script, but rather be read in together with the numerical data.